keps/sig-cli/3517-kubectl-client-side-apply-migration/README.md

```diff
@@ -0,0 +1,1056 @@
+metadata:
+  annotations:
+    kubectl.kubernetes.io/last-applied-configuration: |
+      {"apiVersion":"v1","data":{"key":"value","legacy":"unused"},"kind":"ConfigMap","metadata":{"name":"test","namespace":"default"}}
+  managedFields:
+  - apiVersion: v1
+    fieldsType: FieldsV1
+    fieldsV1:
+      f:data:
+        f:key: {}
+        f:legacy: {}
+    manager: kubectl
+    operation: Apply
+  - apiVersion: v1
+    fieldsType: FieldsV1
+    fieldsV1:
+      f:data:
+        .: {}
+        f:key: {}
+        f:legacy: {}
+      f:metadata:
+        f:annotations:
+          .: {}
+          f:kubectl.kubernetes.io/last-applied-configuration: {}
+    manager: kubectl-client-side-apply
+    operation: Update
+  name: test
+  namespace: default
+```
+
+Note that after the `--server-side` command there are now two owners of `key` and `legacy` fields. 
+This is what causes problems for users. When the user removes the
+field `legacy` via server-side-apply, it surprisingly persists within the object:
+
+
+After field is removed from SSA apply:
+```shell
+cat <<EOF | kubectl apply --server-side -f -                                                                                           csa-to-ssa*
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  key: value
+EOF
+kubectl get configmap test -o yaml --show-managed-fields
+```
+
+```console
+apiVersion: v1
+kind: ConfigMap
+data:
+  key: value
+  legacy: unused
+metadata:
+  annotations:
+    kubectl.kubernetes.io/last-applied-configuration: |
+      {"apiVersion":"v1","data":{"key":"value"},"kind":"ConfigMap","metadata":{"name":"test","namespace":"default"}}
+  managedFields:
+  - apiVersion: v1
+    fieldsType: FieldsV1
+    fieldsV1:
+      f:data:
+        f:key: {}
+    manager: kubectl
+    operation: Apply
+  - apiVersion: v1
+    fieldsType: FieldsV1
+    fieldsV1:
+      f:data:
+        .: {}
+        f:key: {}
+        f:legacy: {}
+      f:metadata:
+        f:annotations:
+          .: {}
+          f:kubectl.kubernetes.io/last-applied-configuration: {}
+    manager: kubectl-client-side-apply
+    operation: Update
+  name: test
+  namespace: default
+```
+
+Note that the `f:legacy: {}` entry was removed from the `kubectl` `Apply` manager,
+but it persists on the `kubectl-client-side-apply` manager. Because the field is
+still owned by at least one manager, it is not removed from the underlying object.
+
+If the user was to now run `kubectl migrate pod test`, the expected fields would
+look like the following:
+
+```console
+apiVersion: v1
+kind: ConfigMap
+data:
+  key: value
+  legacy: unused
+metadata:
+  managedFields:
+  - apiVersion: v1
+    fieldsType: FieldsV1
+    fieldsV1:
+      f:data:
+        .: {}
+        f:key: {}
+        f:legacy: {}
+    manager: kubectl
+    operation: Apply
+  name: test
+  namespace: default
+```
+
+The fields of `kubectl` and `kubectl-client-side-apply` have been unioned and merged
+into the single `Apply` entry of `kubectl`. 
```

Comment:
As we discussed in today's SIG CLI meeting, we'll need to make sure we don't break `apply --prune` functionality with this KEP. Although it is marked as alpha in documentation and has its issues, pruning has been around for a _very_ long time, and plenty of projects and users rely on it (for example, my company's [krane](https://github.com/Shopify/krane)). Here's a TL;DR of how pruning works for the purposes of this discussion:
1. It goes over the GVKs provided in the `--prune-whitelist` argument, or the default list, and makes requests to the API server to accumulate the list of objects to consider (also subject to namespace and label constraints and bugs not relevant to our purposes here).
2. It uses the last-applied annotation to select resources it considers "owned" by kubectl apply: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/apply/prune.go#L116.
3. For any of the "owned" resources that exist server-side but were not included in the current apply, it issues deletion requests.

From a user experience POV, it makes `apply` work declaratively on sets rather than only within objects. IIUC this is still working fine today with both CSA and SSA because the last-applied annotation is being kept in sync (prune doesn't care that the details remain correct, just that it remains present).

We need to make sure that the same set of resources would be pruned before and after migration. Scenarios to consider:
- Some resources kubectl will retrieve in step 1 above may be targeted by SSA operations coming from other actors. For example, a controller may be creating objects (or even just updating a single field within them) using SSA in the same namespace as objects created by a pruning-enabled CD pipeline we want to migrate from CSA to SSA. The CD pipeline must not end up deleting resources created by the controller upon migration. Therefore the presence of `managedFields` itself is definitely not a sufficient signal of pruning scope.
- Pruning needs to work correctly when used in the transitional SSA operation. During that operation, some of the objects we should target may have the last-applied annotation but not any managedField data. Therefore, we need to look at both the new signal _and_ the last-applied annotation.
- With CSA, there was only one possible field manager, so multiple actual actors might be using it. As long as only one of them has pruning enabled, the effect is that that one cleans up after the others. (This doesn't sound like a great idea, but in general we don't use judgements like that in our decisions not to break workflows that have been working for a long time.) This means it's not only the transitional operation itself, but a longer transition period, that needs to look at both the old and the new signal.
- Similar to the above, a single actor might do `apply [subset]`, `apply [subset]`, `apply --prune [full set]` (Krane does this for sequence control purposes). I haven't thought of any particular implications for this one though.
- Conceptually, it is a kinda weird to use the heuristic of "the current manager owns at least one field" to infer that the current manager owns the object for pruning purposes. Therefore, we might want to consider migrating to a bespoke annotation or label instead of the managed fields entry, even if we're still just looking for a match on manager name there. This would also require a transition period.

Reply:
> Therefore, we might want to consider migrating to a bespoke annotation or label instead of the managed fields entry

Does this mean another annotation like "kubectl-pruning-set"-ish like we have for the last applied?
If so, we probably should introduce this earlier than later, since as you mentioned we would need a transition period to make sure that people have had time to apply that. Also, especially for now, we can make the "auto-migration" only on demand behind an experimental flag so that people wouldn't do that, and we could also make sure we make it incompatible with pruning.

I'm trying to see if we can also keep the last applied annotation in the migration process, just for the sake of pruning (that'd would be simpler and wouldn't require a transition period). I think we can still detect a migration if CSA is performed and there is no client-side apply manager in the object.

Sorry, my answer is definitely not super clear, I'm mostly thinking out loud, we can probably make this work, thanks for the write up Katrina!

In the mean time, we're probably going to add TODOs to the KEP and try to get it merged as a draft earlier than later so that we can keep track of what's left to be done, thanks again.

Reply:
> Therefore, we might want to consider migrating to a bespoke annotation or label instead of the managed fields entry

> With CSA, there was only one possible field manager, so multiple actual actors might be using it. As long as only one of them has pruning enabled, the effect is that that one cleans up after the others. (This doesn't sound like a great idea, but in general we don't use judgements like that in our decisions not to break workflows that have been working for a long time.) 

I think a drawback to requiring a special annotation using `--prune` is that there needs to be an initial step that adds the annotation. I think this would break the workflow of multiple actors creating objects, but only one using prune to clean up for the others you also bring up. Therefore I propose the below.

> The CD pipeline must not end up deleting resources created by the controller upon migration. Therefore the presence of managedFields itself is definitely not a sufficient signal of pruning scope.

This is true. Similarly to this thought, recall that kubectl by default provides a field manager of `kubectl`. I propose that for `--prune`'s judgement of "this resource is being declaratively managed by kubectl". , it checks the presence of `kubectl` as an owner of any single field in the object. 

During the transition period the condition can be:

contains last-applied annotation OR `kubectl` appears as a FieldManager in `meta.managedfields`

This method seems closest to today's semantics to me. An object created by a controller would not have `kubectl` in its managed fields anywhere.

It is true that users may choose to use a special field manager on the command line using `--fieldmanager`. To me the best way to handle that would be for prune to also respect the `--fieldmanager` argument when SELECTing resources. I also think in that case it is fair to only consider the `meta.managedFields` and not the last-applied annotation, since the user explicitly expects to use a field manager.

Reply:
> Also, especially for now, we can make the "auto-migration" only on demand behind an experimental flag so that people wouldn't do that, and we could also make sure we make it incompatible with pruning.

By "make it incompatible" do you mean at the flag level, i.e. throw an error if any pruning-related flags are also specified? We should definitely do that if we can't have a compatibility strategy in place, but I'd certainly rather we actually make it work. For one, that would only protect the initial migration, and users would need to read properly and understand that pruning becomes incompatible with apply itself after the migration as well (which we can't reliably detect afaik).

> I'm trying to see if we can also keep the last applied annotation in the migration process, just for the sake of pruning (that'd would be simpler and wouldn't require a transition period). 

That would certainly be the simplest stop-gap solution, allowing us to split out the pruning discussion. Any update on whether that would be workable?

> Does this mean another annotation like "kubectl-pruning-set"-ish like we have for the last applied?
If so, we probably should introduce this earlier than later, since as you mentioned we would need a transition period to make sure that people have had time to apply that. 

Yes, that's what I'm thinking. Since the information we need is the name of the manager, and since its use is object filtering, I'm wondering if we should make it a label instead of an annotation.

> I think we can still detect a migration if CSA is performed and there is no client-side apply manager in the object.

Wouldn't objects created another way and being transitioned to CSA look the same? We emit a warning when users do this, but since we then make it Just Work, I'm sure people do it in the wild.

> I think a drawback to requiring a special annotation using --prune is that there needs to be an initial step that adds the annotation. I think this would break the workflow of multiple actors creating objects, but only one using prune to clean up for the others you also bring up.

Isn't a transition period required regardless? Either we need to have a period where we look at both the old annotation and the field manager, or at both the old annotation and the new one. 

> I also think in that case it is fair to only consider the meta.managedFields and not the last-applied annotation, since the user explicitly expects to use a field manager.

I'd expect the same transition cases to apply regardless of whether they are transitioning to the default field manager or a custom one. We still need a way to do the right thing on the transitional apply operation, and when there may be multiple actors being transitioned.

> I propose that for --prune's judgement of "this resource is being declaratively managed by kubectl". , it checks the presence of kubectl as an owner of any single field in the object.

Although this will make the behaviour the same, IMO it makes less conceptual sense than it used to. Before, there was a single annotation that indicated apply was managing the object, so it made some amount of sense to piggy back on that to have prune manage the object on that basis too. But the field manager conceptually indicates ownership of individual fields only. So it is a much wider conceptual leap to say that a manager owning any fields should mean it also owns the object for pruning purposes. It would now be possible to have conflicts, but for pruning they would be invisible/implicit. I feel like we never would have designed it that way in the first place. In contrast, decoupling to a dedicated label/annotation would _improve_ the conceptual clarity here, making it easier to understand (and even possible for users to intervene on?) what is in scope for pruning by a given manager. In theory, it could even be a step towards splitting pruning out of apply (if we wanted to) or having object ownership conflict detection.

OTOH if we were to separate out a pruning annotation/label, there are some complications to consider. To match current behaviour, we'd need to add it on all apply operations, regardless of pruning-related flag presence. That not only feels weird, but also makes this buggy alpha more visible than it is today, which is not great. And if we allow multiple values for it, we'll need to have an ownership conflict strategy (arguably a good thing, but not something we've had to deal with so far).

In short, CSA removal work is looking like a forcing function to revisit pruning, but since there's a lot to consider there, ideally it would be a separate KEP (I would be happy to get this started if we want). Unfortunately it seems to me that the only option that lets us entirely defer is if it works to retain the existing annotation.

Reply:
> In short, CSA removal work is looking like a forcing function to revisit pruning, but since there's a lot to consider there, ideally it would be a separate KEP (I would be happy to get this started if we want). Unfortunately it seems to me that the only option that lets us entirely defer is if it works to retain the existing annotation.

Yeah, since the update to `prune` seems to be a major blocker for our change, I think it'd be best for us to get started on the KEP for updating it as soon as possible. Really appreciate your offer to get the ball rolling, I'll sync with you offline to see if there is any way I can help.

Reply:
marking this as resolved since it is now a non-goal to remove the annotation